// vite-plugin-top-level-await/dist/utils/make-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAwaitExpression = exports.makeExportListDeclaration = exports.makeMemberExpression = exports.makeReturnStatement = exports.makeCallExpression = exports.makeParenthesisExpression = exports.makeArrowFunction = exports.makeTryCatchStatement = exports.makeArrayExpression = exports.makeImportSpecifier = exports.makeAssignmentStatement = exports.makeAssignmentExpression = exports.makeStatement = exports.makeVariableInitDeclaration = exports.makeVariablesDeclaration = exports.makeIdentifier = void 0;
function span() {
    return { start: 0, end: 0, ctxt: 0 };
}
function makeIdentifier(name) {
    return {
        type: "Identifier",
        span: span(),
        value: name,
        optional: false
    };
}
exports.makeIdentifier = makeIdentifier;
function makeVariablesDeclaration(names) {
    if (names.length === 0)
        return null;
    return {
        type: "VariableDeclaration",
        span: span(),
        kind: "let",
        declare: false,
        declarations: names.map(name => ({
            type: "VariableDeclarator",
            span: span(),
            id: makeIdentifier(name),
            init: null,
            definite: false
        }))
    };
}
exports.makeVariablesDeclaration = makeVariablesDeclaration;
function makeVariableInitDeclaration(name, value) {
    return {
        type: "VariableDeclaration",
        span: span(),
        kind: "let",
        declare: false,
        declarations: [
            {
                type: "VariableDeclarator",
                span: span(),
                id: makeIdentifier(name),
                init: value,
                definite: false
            }
        ]
    };
}
exports.makeVariableInitDeclaration = makeVariableInitDeclaration;
function makeStatement(expression) {
    return {
        type: "ExpressionStatement",
        span: span(),
        expression
    };
}
exports.makeStatement = makeStatement;
function makeAssignmentExpression(left, right) {
    return {
        type: "AssignmentExpression",
        span: span(),
        operator: "=",
        left,
        right
    };
}
exports.makeAssignmentExpression = makeAssignmentExpression;
function makeAssignmentStatement(left, right) {
    const assignmentExpression = makeAssignmentExpression(left, right);
    return makeStatement(left.type === "ObjectPattern" ? makeParenthesisExpression(assignmentExpression) : assignmentExpression);
}
exports.makeAssignmentStatement = makeAssignmentStatement;
function makeImportSpecifier(name, as) {
    return {
        type: "ImportSpecifier",
        span: span(),
        local: makeIdentifier(as),
        imported: as === name ? /* istanbul ignore next */ null : makeIdentifier(name),
        isTypeOnly: false
    };
}
exports.makeImportSpecifier = makeImportSpecifier;
function makeArrayExpression(items) {
    return {
        type: "ArrayExpression",
        span: span(),
        elements: items.map(item => ({
            spread: null,
            expression: item
        }))
    };
}
exports.makeArrayExpression = makeArrayExpression;
function makeTryCatchStatement(tryStatements, catchStatements) {
    return {
        type: "TryStatement",
        span: span(),
        block: {
            type: "BlockStatement",
            span: span(),
            stmts: tryStatements
        },
        handler: {
            type: "CatchClause",
            span: span(),
            param: null,
            body: {
                type: "BlockStatement",
                span: span(),
                stmts: catchStatements
            }
        },
        finalizer: null
    };
}
exports.makeTryCatchStatement = makeTryCatchStatement;
function makeArrowFunction(args, statements, async) {
    return {
        type: "ArrowFunctionExpression",
        span: span(),
        params: args.map(arg => ({
            type: "Identifier",
            span: span(),
            value: arg,
            optional: false
        })),
        body: {
            type: "BlockStatement",
            span: span(),
            stmts: statements
        },
        async: !!async,
        generator: false,
        typeParameters: null,
        returnType: null
    };
}
exports.makeArrowFunction = makeArrowFunction;
function makeParenthesisExpression(expression) {
    return {
        type: "ParenthesisExpression",
        span: span(),
        expression
    };
}
exports.makeParenthesisExpression = makeParenthesisExpression;
function makeCallExpression(functionExpression, args) {
    return {
        type: "CallExpression",
        span: span(),
        // Put IIFE's function expression in (parenthesis)
        callee: functionExpression.type === "FunctionExpression" || functionExpression.type === "ArrowFunctionExpression"
            ? makeParenthesisExpression(functionExpression)
            : functionExpression,
        arguments: (args !== null && args !== void 0 ? args : []).map(arg => ({
            spread: null,
            expression: arg
        })),
        typeArguments: null
    };
}
exports.makeCallExpression = makeCallExpression;
function makeReturnStatement(expression) {
    return {
        type: "ReturnStatement",
        span: span(),
        argument: expression
    };
}
exports.makeReturnStatement = makeReturnStatement;
function makeMemberExpression(object, member) {
    return {
        type: "MemberExpression",
        span: span(),
        object: typeof object === "string" ? makeIdentifier(object) : object,
        property: makeIdentifier(member)
    };
}
exports.makeMemberExpression = makeMemberExpression;
function makeExportListDeclaration(map) {
    return {
        type: "ExportNamedDeclaration",
        span: span(),
        specifiers: map.map(([exportName, identifier]) => ({
            type: "ExportSpecifier",
            span: span(),
            orig: makeIdentifier(identifier),
            exported: identifier === exportName ? null : makeIdentifier(exportName),
            isTypeOnly: false
        })),
        source: null,
        // @ts-ignore
        typeOnly: false,
        // @ts-ignore
        assets: null
    };
}
exports.makeExportListDeclaration = makeExportListDeclaration;
function makeAwaitExpression(expression) {
    return {
        type: "AwaitExpression",
        span: span(),
        argument: expression
    };
}
exports.makeAwaitExpression = makeAwaitExpression;
