// vite-plugin-vercel/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => allPlugins
});
module.exports = __toCommonJS(src_exports);
var import_promises4 = __toESM(require("fs/promises"), 1);

// src/utils.ts
var import_vite = require("vite");
var import_path = __toESM(require("path"), 1);
function getRoot(config) {
  return (0, import_vite.normalizePath)(config.root || process.cwd());
}
function getOutput(config, suffix) {
  var _a, _b;
  return import_path.default.join(
    ((_a = config.vercel) == null ? void 0 : _a.outDir) ? "" : getRoot(config),
    ((_b = config.vercel) == null ? void 0 : _b.outDir) ?? ".vercel/output",
    suffix ?? ""
  );
}
function getPublic(config) {
  return import_path.default.join(getRoot(config), config.publicDir || "public");
}
function pathRelativeTo(filePath, config, rel) {
  const root = getRoot(config);
  return (0, import_vite.normalizePath)(
    import_path.default.relative((0, import_vite.normalizePath)(import_path.default.join(root, rel)), filePath)
  );
}

// src/config.ts
var import_path2 = __toESM(require("path"), 1);

// src/schemas/config/config.ts
var import_zod = require("zod");
var HasOrMissing = import_zod.z.array(
  import_zod.z.union([
    import_zod.z.object({
      type: import_zod.z.literal("host"),
      value: import_zod.z.string()
    }).strict(),
    import_zod.z.object({
      type: import_zod.z.literal("header"),
      key: import_zod.z.string(),
      value: import_zod.z.string().optional()
    }).strict(),
    import_zod.z.object({
      type: import_zod.z.literal("cookie"),
      key: import_zod.z.string(),
      value: import_zod.z.string().optional()
    }).strict(),
    import_zod.z.object({
      type: import_zod.z.literal("query"),
      key: import_zod.z.string(),
      value: import_zod.z.string().optional()
    }).strict()
  ])
).optional();
var vercelOutputConfigSchema = import_zod.z.object({
  version: import_zod.z.literal(3),
  routes: import_zod.z.array(
    import_zod.z.union([
      import_zod.z.object({
        src: import_zod.z.string(),
        dest: import_zod.z.string().optional(),
        headers: import_zod.z.record(import_zod.z.string()).optional(),
        methods: import_zod.z.array(import_zod.z.string()).optional(),
        status: import_zod.z.number().int().positive().optional(),
        continue: import_zod.z.boolean().optional(),
        check: import_zod.z.boolean().optional(),
        missing: HasOrMissing,
        has: HasOrMissing,
        locale: import_zod.z.object({
          redirect: import_zod.z.record(import_zod.z.string()).optional(),
          cookie: import_zod.z.string().optional()
        }).strict().optional(),
        middlewarePath: import_zod.z.string().optional()
      }).strict(),
      import_zod.z.object({
        handle: import_zod.z.union([
          import_zod.z.literal("rewrite"),
          import_zod.z.literal("filesystem"),
          import_zod.z.literal("resource"),
          import_zod.z.literal("miss"),
          import_zod.z.literal("hit"),
          import_zod.z.literal("error")
        ]),
        src: import_zod.z.string().optional(),
        dest: import_zod.z.string().optional(),
        status: import_zod.z.number().optional()
      }).strict()
    ])
  ).optional(),
  images: import_zod.z.object({
    sizes: import_zod.z.tuple([
      import_zod.z.number().int().positive(),
      import_zod.z.number().int().positive()
    ]),
    domains: import_zod.z.array(import_zod.z.string()).nonempty().optional(),
    minimumCacheTTL: import_zod.z.number().int().positive().optional(),
    formats: import_zod.z.union([import_zod.z.literal("image/avif"), import_zod.z.literal("image/webp")]).array().nonempty().optional(),
    dangerouslyAllowSVG: import_zod.z.boolean().optional(),
    contentSecurityPolicy: import_zod.z.string().optional()
  }).strict().optional(),
  wildcard: import_zod.z.array(
    import_zod.z.object({
      domain: import_zod.z.string(),
      value: import_zod.z.string()
    }).strict()
  ).optional(),
  overrides: import_zod.z.record(
    import_zod.z.object({
      path: import_zod.z.string().optional(),
      contentType: import_zod.z.string().optional()
    }).strict()
  ).optional(),
  cache: import_zod.z.array(import_zod.z.string()).optional()
}).strict();

// src/config.ts
var import_promises = __toESM(require("fs/promises"), 1);
var import_routing_utils = require("@vercel/routing-utils");
function reorderEnforce(arr) {
  return [
    ...arr.filter((r) => r.enforce === "pre"),
    ...arr.filter((r) => !r.enforce),
    ...arr.filter((r) => r.enforce === "post")
  ];
}
function getConfig(resolvedConfig, rewrites, overrides) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l;
  const _rewrites = [
    // User provided config always comes first
    ...((_a = resolvedConfig.vercel) == null ? void 0 : _a.rewrites) ?? [],
    ...rewrites ?? []
  ];
  const { routes, error } = (0, import_routing_utils.getTransformedRoutes)({
    cleanUrls: ((_b = resolvedConfig.vercel) == null ? void 0 : _b.cleanUrls) ?? true,
    trailingSlash: (_c = resolvedConfig.vercel) == null ? void 0 : _c.trailingSlash,
    rewrites: reorderEnforce(_rewrites),
    redirects: ((_d = resolvedConfig.vercel) == null ? void 0 : _d.redirects) ? reorderEnforce((_e = resolvedConfig.vercel) == null ? void 0 : _e.redirects) : void 0
  });
  if (error) {
    throw error;
  }
  if (((_g = (_f = resolvedConfig.vercel) == null ? void 0 : _f.config) == null ? void 0 : _g.routes) && resolvedConfig.vercel.config.routes.length > 0) {
    console.warn(
      "It is discouraged to use `vercel.config.routes` to override routes. Prefer using `vercel.rewrites` and `vercel.redirects`."
    );
  }
  return vercelOutputConfigSchema.parse({
    version: 3,
    ...(_h = resolvedConfig.vercel) == null ? void 0 : _h.config,
    routes: [
      ...routes ?? [],
      ...((_j = (_i = resolvedConfig.vercel) == null ? void 0 : _i.config) == null ? void 0 : _j.routes) ?? []
    ],
    overrides: {
      ...(_l = (_k = resolvedConfig.vercel) == null ? void 0 : _k.config) == null ? void 0 : _l.overrides,
      ...overrides
    }
  });
}
function getConfigDestination(resolvedConfig) {
  return import_path2.default.join(getOutput(resolvedConfig), "config.json");
}
async function writeConfig(resolvedConfig, rewrites, overrides) {
  await import_promises.default.writeFile(
    getConfigDestination(resolvedConfig),
    JSON.stringify(
      getConfig(resolvedConfig, rewrites, overrides),
      void 0,
      2
    ),
    "utf-8"
  );
}

// src/build.ts
var import_fast_glob = __toESM(require("fast-glob"), 1);
var import_path3 = __toESM(require("path"), 1);
var import_esbuild = require("esbuild");

// src/assert.ts
var import_libassert = require("@brillout/libassert");
var libName = "vite-plugin-vercel";
function assert(condition, errorMessage) {
  if (condition) {
    return;
  }
  const err = (0, import_libassert.newError)(`[${libName}][Wrong Usage] ${errorMessage}`, 2);
  throw err;
}

// src/schemas/config/vc-config.ts
var import_zod2 = require("zod");
var vercelOutputEdgeVcConfigSchema = import_zod2.z.object({
  runtime: import_zod2.z.literal("edge"),
  entrypoint: import_zod2.z.string(),
  envVarsInUse: import_zod2.z.array(import_zod2.z.string()).optional()
}).strict();
var vercelOutputServerlessVcConfigSchema = import_zod2.z.object({
  runtime: import_zod2.z.string(),
  handler: import_zod2.z.string(),
  memory: import_zod2.z.number().int().min(128).max(3008).optional(),
  maxDuration: import_zod2.z.number().int().positive().optional(),
  environment: import_zod2.z.record(import_zod2.z.string()).optional(),
  regions: import_zod2.z.array(import_zod2.z.string()).optional()
}).strict();
var vercelOutputServerlessNodeVcConfigSchema = vercelOutputServerlessVcConfigSchema.extend({
  launcherType: import_zod2.z.literal("Nodejs"),
  shouldAddHelpers: import_zod2.z.boolean().optional(),
  shouldAddSourcemapSupport: import_zod2.z.boolean().optional(),
  awsLambdaHandler: import_zod2.z.string().optional()
}).strict();
var vercelOutputVcConfigSchema = import_zod2.z.union([
  vercelOutputEdgeVcConfigSchema,
  vercelOutputServerlessVcConfigSchema,
  vercelOutputServerlessNodeVcConfigSchema
]);

// src/build.ts
var import_promises2 = __toESM(require("fs/promises"), 1);
function getAdditionalEndpoints(resolvedConfig) {
  var _a;
  return (((_a = resolvedConfig.vercel) == null ? void 0 : _a.additionalEndpoints) ?? []).map((e) => ({
    ...e,
    addRoute: e.addRoute ?? true,
    // path.resolve removes the trailing slash if any
    destination: import_path3.default.posix.resolve("/", e.destination) + ".func"
  }));
}
function getEntries(resolvedConfig) {
  const apiEntries = import_fast_glob.default.sync(`${getRoot(resolvedConfig)}/api/**/*.*([a-zA-Z0-9])`).filter((filepath) => !import_path3.default.basename(filepath).startsWith("_"));
  if (apiEntries.length > 0) {
    console.warn(
      "@vercel/build is currently force building /api files itself, with no way to disable it. In order to avoid double compilation, you should temporarily rename /api to /_api while using this plugin. /_api functions are compiled under .vercel/output/functions/api/*.func as if they were in /api."
    );
  }
  const otherApiEntries = import_fast_glob.default.sync(`${getRoot(resolvedConfig)}/_api/**/*.*([a-zA-Z0-9])`).filter((filepath) => !import_path3.default.basename(filepath).startsWith("_"));
  return [...apiEntries, ...otherApiEntries].reduce((entryPoints, filePath) => {
    const outFilePath = pathRelativeTo(
      filePath,
      resolvedConfig,
      filePath.includes("/_api/") ? "_api" : "api"
    );
    const parsed = import_path3.default.posix.parse(outFilePath);
    entryPoints.push({
      source: filePath,
      destination: `api/${import_path3.default.posix.join(parsed.dir, parsed.name)}.func`,
      addRoute: true
    });
    return entryPoints;
  }, getAdditionalEndpoints(resolvedConfig));
}
var standardBuildOptions = {
  bundle: true,
  target: "es2020",
  format: "cjs",
  platform: "node",
  logLevel: "info",
  minify: true
};
async function buildFn(resolvedConfig, entry, buildOptions) {
  assert(
    entry.destination.length > 0,
    `Endpoint ${typeof entry.source === "string" ? entry.source : "-"} does not have build destination`
  );
  const outfile = import_path3.default.join(
    getOutput(resolvedConfig, "functions"),
    entry.destination,
    "index.js"
  );
  const options = Object.assign({}, standardBuildOptions, { outfile });
  if (buildOptions) {
    Object.assign(options, buildOptions);
  }
  if (!options.stdin) {
    if (typeof entry.source === "string") {
      options.entryPoints = [entry.source];
    } else {
      assert(
        typeof entry.source === "object",
        `\`{ source }\` must be a string or an object`
      );
      assert(
        typeof entry.source.contents === "string",
        `\`{ contents }\` must be a string`
      );
      options.stdin = entry.source;
    }
  }
  await (0, import_esbuild.build)(options);
  await writeVcConfig(resolvedConfig, entry.destination, Boolean(entry.edge));
}
async function writeVcConfig(resolvedConfig, destination, edge) {
  var _a;
  const vcConfig = import_path3.default.join(
    getOutput(resolvedConfig, "functions"),
    destination,
    ".vc-config.json"
  );
  await import_promises2.default.writeFile(
    vcConfig,
    JSON.stringify(
      vercelOutputVcConfigSchema.parse(
        edge ? {
          runtime: "edge",
          entrypoint: "index.js"
        } : {
          runtime: "nodejs16.x",
          handler: "index.js",
          maxDuration: (_a = resolvedConfig.vercel) == null ? void 0 : _a.defaultMaxDuration,
          launcherType: "Nodejs",
          shouldAddHelpers: true
        }
      ),
      void 0,
      2
    ),
    "utf-8"
  );
}
function getSourceAndDestination(destination) {
  if (destination.startsWith("api/")) {
    return import_path3.default.posix.resolve("/", destination);
  }
  return import_path3.default.posix.resolve("/", destination, ":match*");
}
async function buildEndpoints(resolvedConfig) {
  const entries = getEntries(resolvedConfig);
  for (const entry of entries) {
    await buildFn(resolvedConfig, entry);
  }
  return entries.filter((e) => e.addRoute !== false).map((e) => e.destination.replace(/\.func$/, "")).map((destination) => ({
    source: getSourceAndDestination(destination),
    destination: getSourceAndDestination(destination)
  }));
}

// src/prerender.ts
var import_path4 = __toESM(require("path"), 1);

// src/schemas/config/prerender-config.ts
var import_zod3 = require("zod");
var vercelOutputPrerenderConfigSchema = import_zod3.z.object({
  expiration: import_zod3.z.union([import_zod3.z.number().int().positive(), import_zod3.z.literal(false)]),
  group: import_zod3.z.number().int().optional(),
  bypassToken: import_zod3.z.string().optional(),
  fallback: import_zod3.z.string().optional(),
  allowQuery: import_zod3.z.array(import_zod3.z.string()).optional()
}).strict();

// src/prerender.ts
var import_promises3 = __toESM(require("fs/promises"), 1);
function execPrerender(resolvedConfig) {
  var _a;
  const prerender = (_a = resolvedConfig.vercel) == null ? void 0 : _a.prerender;
  if (prerender === false) {
    return;
  }
  return prerender == null ? void 0 : prerender(resolvedConfig);
}
var group = 1;
async function writePrerenderConfig(resolvedConfig, destination, isr) {
  const parsed = import_path4.default.parse(destination);
  const outfile = import_path4.default.join(
    getOutput(resolvedConfig, "functions"),
    parsed.dir,
    parsed.name + ".prerender-config.json"
  );
  await import_promises3.default.mkdir(
    import_path4.default.join(getOutput(resolvedConfig, "functions"), parsed.dir),
    { recursive: true }
  );
  await import_promises3.default.writeFile(
    outfile,
    JSON.stringify(
      vercelOutputPrerenderConfigSchema.parse({
        group: group++,
        ...isr
      }),
      void 0,
      2
    ),
    "utf-8"
  );
}
function getPrerenderSymlinkInfo(resolvedConfig, destination, target) {
  const parsed = import_path4.default.parse(destination);
  const targetParsed = import_path4.default.parse(target);
  return {
    target: import_path4.default.join(
      getOutput(resolvedConfig, "functions"),
      targetParsed.dir,
      targetParsed.name + ".func"
    ),
    link: import_path4.default.join(
      getOutput(resolvedConfig, "functions"),
      parsed.dir,
      parsed.name + ".func"
    )
  };
}
async function buildPrerenderConfigs(resolvedConfig) {
  const isr = await getIsrConfig(resolvedConfig);
  const entries = Object.entries(isr);
  const rewrites = [];
  for (const [destination, { symlink, route, ...isr2 }] of entries) {
    await writePrerenderConfig(resolvedConfig, destination, isr2);
    if (symlink) {
      const info = getPrerenderSymlinkInfo(
        resolvedConfig,
        destination,
        symlink
      );
      await copyDir(info.target, info.link);
    }
    if (route) {
      rewrites.push({
        source: `(${route})`,
        destination: `${destination}/?__original_path=$1`
      });
    }
  }
  return rewrites;
}
async function copyDir(src, dest) {
  await import_promises3.default.mkdir(dest, { recursive: true });
  const entries = await import_promises3.default.readdir(src, { withFileTypes: true });
  for (const entry of entries) {
    const srcPath = import_path4.default.join(src, entry.name);
    const destPath = import_path4.default.join(dest, entry.name);
    entry.isDirectory() ? await copyDir(srcPath, destPath) : await import_promises3.default.copyFile(srcPath, destPath);
  }
}
async function getIsrConfig(resolvedConfig) {
  var _a;
  const isr = ((_a = resolvedConfig.vercel) == null ? void 0 : _a.isr) ?? {};
  if (typeof isr === "function") {
    return await isr();
  }
  return isr;
}

// src/index.ts
var import_path5 = __toESM(require("path"), 1);
function vercelPlugin() {
  let resolvedConfig;
  return {
    apply: "build",
    name: "vite-plugin-vercel",
    enforce: "post",
    configResolved(config) {
      resolvedConfig = config;
    },
    async buildStart() {
      if (process.env.VERCEL_ENV === "production" && !process.env.ENABLE_VC_BUILD) {
        throw new Error(
          "Missing ENABLE_VC_BUILD=1 to your environment variables in your project settings"
        );
      }
    },
    async writeBundle() {
      var _a;
      if (!((_a = resolvedConfig.build) == null ? void 0 : _a.ssr)) {
        await cleanOutputDirectory(resolvedConfig);
        return;
      }
      const overrides = await execPrerender(resolvedConfig);
      const userOverrides = await computeStaticHtmlOverrides(resolvedConfig);
      const rewrites = await buildEndpoints(resolvedConfig);
      rewrites.push(...await buildPrerenderConfigs(resolvedConfig));
      await writeConfig(resolvedConfig, rewrites, {
        ...userOverrides,
        ...overrides
      });
    }
  };
}
async function cleanOutputDirectory(resolvedConfig) {
  await import_promises4.default.rm(getOutput(resolvedConfig), {
    recursive: true,
    force: true
  });
}
async function computeStaticHtmlOverrides(resolvedConfig) {
  const staticAbsolutePath = getOutput(resolvedConfig, "static");
  const files = await getStaticHtmlFiles(staticAbsolutePath);
  const publicDir = getPublic(resolvedConfig);
  const publicFiles = await getStaticHtmlFiles(publicDir);
  files.push(
    ...publicFiles.map((f) => f.replace(publicDir, staticAbsolutePath))
  );
  return files.reduce((acc, curr) => {
    const relPath = import_path5.default.relative(staticAbsolutePath, curr);
    const parsed = import_path5.default.parse(relPath);
    const pathJoined = import_path5.default.join(parsed.dir, parsed.name);
    acc[relPath] = {
      path: pathJoined
    };
    return acc;
  }, {});
}
async function getStaticHtmlFiles(src) {
  try {
    await import_promises4.default.stat(src);
  } catch (e) {
    return [];
  }
  const entries = await import_promises4.default.readdir(src, { withFileTypes: true });
  const htmlFiles = [];
  for (const entry of entries) {
    const srcPath = import_path5.default.join(src, entry.name);
    entry.isDirectory() ? htmlFiles.push(...await getStaticHtmlFiles(srcPath)) : srcPath.endsWith(".html") ? htmlFiles.push(srcPath) : void 0;
  }
  return htmlFiles;
}
function allPlugins() {
  return [vercelPlugin()];
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
